src/EditorFeatures/Core/Tagging/AsynchronousViewportTaggerProvider.SingleViewportTaggerProvider.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Collections;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Workspaces;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Tagging;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Tagging;
+
+internal abstract partial class AsynchronousViewportTaggerProvider<TTag> where TTag : ITag
+{
+    private sealed class SingleViewportTaggerProvider : AsynchronousViewTaggerProvider<TTag>
+    {
+        private readonly AsynchronousViewportTaggerProvider<TTag> _callback;
+
+        private readonly ViewPortToTag _viewPortToTag;
+
+        public SingleViewportTaggerProvider(
+            AsynchronousViewportTaggerProvider<TTag> callback,
+            ViewPortToTag viewPortToTag,
+            IThreadingContext threadingContext,
+            IGlobalOptionService globalOptions,
+            ITextBufferVisibilityTracker? visibilityTracker,
+            IAsynchronousOperationListener asyncListener)
+            : base(threadingContext, globalOptions, visibilityTracker, asyncListener)
+        {
+            _callback = callback;
+            _viewPortToTag = viewPortToTag;
+        }
+
+        protected override ImmutableArray<IOption2> Options
+            => _callback.Options;
+
+        protected override TaggerTextChangeBehavior TextChangeBehavior
+            => _callback.TextChangeBehavior;
+
+        protected override ITaggerEventSource CreateEventSource(ITextView textView, ITextBuffer subjectBuffer)
+            => _callback.CreateEventSource(textView, subjectBuffer);
+
+        protected override TaggerDelay EventChangeDelay
+            // If we're in view, tag at the cadence the feature wants for visible code.  Otherwise, if we're out of
+            // view, tag at the slowest cadence to reduce the amount of resources used for things that may never even be
+            // looked at.
+            => _viewPortToTag == ViewPortToTag.InView ? _callback.EventChangeDelay : TaggerDelay.NonFocus;
```

Comment:
core idea differentiating the cadence of the 'in view' tagging vs the 'outside teh view' tagging.